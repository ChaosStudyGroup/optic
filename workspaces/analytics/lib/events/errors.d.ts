export interface JavascriptErrorDetectedInFrontendProperties {
    message: string;
    stack: string;
}
