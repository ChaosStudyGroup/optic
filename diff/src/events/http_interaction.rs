use super::EventLoadingError;
use avro_rs;
use cqrs_core::Event;
use serde::Deserialize;
use serde_json;
use std::io;

// TODO: consider whether these aren't actually Events and the Traverser not an Aggregator

#[derive(Deserialize, Debug)]
pub struct HttpInteraction {
  pub uuid: String,
  pub request: Request,
  pub response: Response,
  pub tags: Vec<HttpInteractionTag>,
}

#[derive(Deserialize, Debug)]
pub struct HttpInteractionTag {
  name: String,
  value: String,
}

#[derive(Deserialize, Debug)]
pub struct Request {
  pub host: String,
  pub method: String,
  pub path: String,
  // #[serde(skip)]
  pub query: ArbitraryData,
  pub body: Body,
}

#[derive(Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct Response {
  status_code: u16,
  // #[serde(skip)]
  headers: ArbitraryData,
}

#[derive(Deserialize, Debug)]
pub struct Body {
  content_type: Option<String>,
  // #[serde(skip)]
  value: ArbitraryData,
}

#[derive(Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct ArbitraryData {
  shape_hash_v1_base64: Option<String>,
  as_json_string: Option<String>,
  as_text: Option<String>,
}

// impl Default for ArbitraryData {
//   fn default() -> Self {
//     Self {}
//   }
// }

impl Event for HttpInteraction {
  fn event_type(&self) -> &'static str {
    "HttpInteraction"
  }
}

impl HttpInteraction {
  pub fn from_json_str(json: &str) -> Result<Self, serde_json::Error> {
    serde_json::from_str(json)
  }

  pub fn from_avro() -> HttpInteractionAvroDeserializer {
    let deserializer = HttpInteractionAvroDeserializer::new(&INTERACTION_AVRO_SCHEMA)
      .expect("interaction avro schema should be valid");

    deserializer
  }
}

pub struct HttpInteractionAvroDeserializer {
  schema: avro_rs::Schema,
}

impl HttpInteractionAvroDeserializer {
  fn new(schema_str: &str) -> Result<Self, avro_rs::Error> {
    let schema = avro_rs::Schema::parse_str(schema_str)?;
    Ok(Self { schema })
  }

  pub fn reader<'a, R>(&'a mut self, source: R) -> Result<avro_rs::Reader<'a, R>, avro_rs::Error>
  where
    R: io::Read,
    R: 'a,
  {
    avro_rs::Reader::with_schema(&self.schema, source)
  }
}

const INTERACTION_AVRO_SCHEMA: &str = r#"{
  "type": "record",
  "name": "InteractionBatch",
  "namespace": "com.useoptic.types.capture",
  "fields": [
    {
      "name": "groupingIdentifiers",
      "type": {
        "type": "record",
        "name": "GroupingIdentifiers",
        "fields": [
          { "name": "agentGroupId", "type": "string" },
          { "name": "captureId", "type": "string" },
          { "name": "agentId", "type": "string" },
          { "name": "batchId", "type": "string" }
        ]
      }
    },
    {
      "name": "batchItems",
      "type": {
        "type": "array",
        "items": {
          "type": "record",
          "name": "HttpInteraction",
          "fields": [
            { "name": "uuid", "type": "string" },
            {
              "name": "request",
              "type": {
                "type": "record",
                "name": "Request",
                "fields": [
                  { "name": "host", "type": "string" },
                  { "name": "method", "type": "string" },
                  { "name": "path", "type": "string" },
                  {
                    "name": "query",
                    "type": {
                      "type": "record",
                      "name": "ArbitraryData",
                      "fields": [
                        {
                          "name": "shapeHashV1Base64",
                          "type": ["null", "string"],
                          "default": null
                        },
                        {
                          "name": "asJsonString",
                          "type": ["null", "string"],
                          "default": null
                        },
                        {
                          "name": "asText",
                          "type": ["null", "string"],
                          "default": null
                        }
                      ]
                    }
                  },
                  {
                    "name": "headers",
                    "type": {
                      "type": "record",
                      "name": "ArbitraryData",
                      "fields": [
                        {
                          "name": "shapeHashV1Base64",
                          "type": ["null", "string"],
                          "default": null
                        },
                        {
                          "name": "asJsonString",
                          "type": ["null", "string"],
                          "default": null
                        },
                        {
                          "name": "asText",
                          "type": ["null", "string"],
                          "default": null
                        }
                      ]
                    }
                  },
                  {
                    "name": "body",
                    "type": {
                      "type": "record",
                      "name": "Body",
                      "fields": [
                        { "name": "contentType", "type": ["null", "string"] },
                        {
                          "name": "value",
                          "type": {
                            "type": "record",
                            "name": "ArbitraryData",
                            "fields": [
                              {
                                "name": "shapeHashV1Base64",
                                "type": ["null", "string"],
                                "default": null
                              },
                              {
                                "name": "asJsonString",
                                "type": ["null", "string"],
                                "default": null
                              },
                              {
                                "name": "asText",
                                "type": ["null", "string"],
                                "default": null
                              }
                            ]
                          }
                        }
                      ]
                    }
                  }
                ]
              }
            },
            {
              "name": "response",
              "type": {
                "type": "record",
                "name": "Response",
                "fields": [
                  { "name": "statusCode", "type": "int" },
                  {
                    "name": "headers",
                    "type": {
                      "type": "record",
                      "name": "ArbitraryData",
                      "fields": [
                        {
                          "name": "shapeHashV1Base64",
                          "type": ["null", "string"],
                          "default": null
                        },
                        {
                          "name": "asJsonString",
                          "type": ["null", "string"],
                          "default": null
                        },
                        {
                          "name": "asText",
                          "type": ["null", "string"],
                          "default": null
                        }
                      ]
                    }
                  },
                  {
                    "name": "body",
                    "type": {
                      "type": "record",
                      "name": "Body",
                      "fields": [
                        {
                          "name": "contentType",
                          "type": ["null", "string"]
                        },
                        {
                          "name": "value",
                          "type": {
                            "type": "record",
                            "name": "ArbitraryData",
                            "fields": [
                              {
                                "name": "shapeHashV1Base64",
                                "type": ["null", "string"],
                                "default": null
                              },
                              {
                                "name": "asJsonString",
                                "type": ["null", "string"],
                                "default": null
                              },
                              {
                                "name": "asText",
                                "type": ["null", "string"],
                                "default": null
                              }
                            ]
                          }
                        }
                      ]
                    }
                  }
                ]
              }
            },
            {
              "name": "tags",
              "type": {
                "type": "array",
                "items": {
                  "type": "record",
                  "name": "HttpInteractionTag",
                  "fields": [
                    { "name": "name", "type": "string" },
                    { "name": "value", "type": "string" }
                  ]
                }
              }
            }
          ]
        }
      }
    }
  ]
}
"#;

#[cfg(test)]
mod test {
  use super::*;
  #[test]
  fn can_deserialize_interaction_from_json_str() {
    let json = r#"{
      "uuid": "3",
      "request": {
        "host": "localhost",
        "method": "GET",
        "path": "/todos",
        "query": {},
        "headers": {
          "asJsonString": null,
          "asText": null,
          "asShapeHashBytes": null
        },
        "body": {
          "contentType": null,
          "value": {}
        }
      },
      "response": {
        "statusCode": 200,
        "headers": {
          "asJsonString": null,
          "asText": null,
          "asShapeHashBytes": null
        },
        "body": {
          "contentType": "application/json",
          "value": {}
        }
      },
      "tags": []
    }"#;

    let interaction = HttpInteraction::from_json_str(&json);
    interaction.expect("Valid JSON should be able to deserialize into an HttpInteraction");
  }
}
