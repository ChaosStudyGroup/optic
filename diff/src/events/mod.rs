#![allow(dead_code)]

use cqrs_core::Event;
use serde::Deserialize;
use serde_json;
use std::fs;
use std::io;

pub mod endpoint;
pub mod http_interaction;
pub mod rfc;
pub mod shape;

pub use endpoint::EndpointEvent;
pub use http_interaction::HttpInteraction;

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EventContext {
  client_id: String,
  client_session_id: String,
  client_command_batch_id: String,
  created_at: String,
}

#[derive(Deserialize)]
#[serde(untagged)]
pub enum SpecEvent {
  EndpointEvent(endpoint::EndpointEvent),
  RfcEvent(rfc::RfcEvent),
  ShapeEvent(shape::ShapeEvent),
}

impl Event for SpecEvent {
  fn event_type(&self) -> &'static str {
    match self {
      SpecEvent::EndpointEvent(evt) => evt.event_type(),
      SpecEvent::RfcEvent(evt) => evt.event_type(),
      SpecEvent::ShapeEvent(evt) => evt.event_type(),
    }
  }
}

impl SpecEvent {
  pub fn from_file(filename: &str) -> Result<Vec<SpecEvent>, EventLoadingError> {
    let file_contents =
      fs::read_to_string(filename).expect(&format!("File at {} could not be read", &filename));

    let events: Vec<SpecEvent> = serde_json::from_str(&file_contents)?;

    Ok(events)
  }
}

#[derive(Debug)]
pub enum EventLoadingError {
  Io(io::Error),
  Json(serde_json::Error),
}

// We only have to implement this trait for serde_json::Error as handle conversion from the io::Error
impl From<serde_json::Error> for EventLoadingError {
  fn from(err: serde_json::Error) -> EventLoadingError {
    use serde_json::error::Category;
    match err.classify() {
      Category::Io => EventLoadingError::Io(err.into()),
      Category::Syntax | Category::Data | Category::Eof => EventLoadingError::Json(err),
    }
  }
}
