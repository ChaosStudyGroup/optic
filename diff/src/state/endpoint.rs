use super::shape::ShapeId;
use serde::Deserialize;

pub type PathComponentId = String;
pub type PathComponentIdRef<'a> = &'a str;
pub type RequestId = String;
pub type RequestParameterId = String;
pub type ResponseId = String;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ShapedBodyDescriptor {
  http_content_type: String,
  shape_id: ShapeId,
  is_removed: bool,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ShapedRequestParameterShapeDescriptor {
  shape_id: ShapeId,
  is_removed: bool,
}
