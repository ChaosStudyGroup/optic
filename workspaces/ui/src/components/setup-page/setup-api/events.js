import keymirror from 'keymirror';

export const DocumentAPIFlowEvents = keymirror({
  MODE_SELECTED: null,
  FRAMEWORK_SELECTED: null,
  EMPTY_SEARCH_RESULTS: null,
  CHANGED_START_COMMAND: null,
  CHANGED_HOSTNAME: null,
  SET_STEP_TO: null,
  CHANGED_TARGET_URL: null,
  MARK_API_AS_INITIALIZED: null,
});

export function MODE_SELECTED(mode) {
  return { type: DocumentAPIFlowEvents.MODE_SELECTED, mode };
}
export function SET_STEP_TO(to) {
  return { type: DocumentAPIFlowEvents.SET_STEP_TO, to };
}
export function FRAMEWORK_SELECTED(framework) {
  return { type: DocumentAPIFlowEvents.FRAMEWORK_SELECTED, framework };
}
export function EMPTY_SEARCH_RESULTS() {
  return { type: DocumentAPIFlowEvents.EMPTY_SEARCH_RESULTS };
}

export function CHANGED_START_COMMAND(command) {
  return { type: DocumentAPIFlowEvents.CHANGED_START_COMMAND, command };
}

export function CHANGED_HOSTNAME(hostname) {
  return { type: DocumentAPIFlowEvents.CHANGED_HOSTNAME, hostname };
}
export function CHANGED_TARGET_URL(targetUrl) {
  return { type: DocumentAPIFlowEvents.CHANGED_TARGET_URL, targetUrl };
}

export function MARK_API_AS_INITIALIZED(cwd, name) {
  return { type: DocumentAPIFlowEvents.MARK_API_AS_INITIALIZED, name, cwd };
}

//constants
export const MODES = keymirror({
  RECOMMENDED: null,
  MANUAL: null,
});
