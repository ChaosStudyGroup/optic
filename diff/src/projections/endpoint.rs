use crate::events::{EndpointEvent, SpecEvent};
use crate::state::endpoint::PathComponentId;
use cqrs_core::{Aggregate, AggregateEvent};
use std::collections::HashMap;

pub struct EndpointProjection {
  pub absolute_paths: HashMap<PathComponentId, String>,
}

impl EndpointProjection {
  pub fn with_path(
    &mut self,
    parent_path_id: PathComponentId,
    path_id: PathComponentId,
    path_name: String,
  ) {
    let absolute_paths = &mut self.absolute_paths;

    let parent_absolute_path = absolute_paths
      .get(&parent_path_id)
      .expect("expected parent_path_id to already exist");
    let absolute_path = format!("{}/{}", &parent_absolute_path, &path_name);
    absolute_paths.insert(path_id, absolute_path);
  }
}

impl Default for EndpointProjection {
  fn default() -> Self {
    let mut absolute_paths = HashMap::new();
    absolute_paths.insert(PathComponentId::from("root"), String::from(""));
    EndpointProjection { absolute_paths }
  }
}

impl Aggregate for EndpointProjection {
  fn aggregate_type() -> &'static str {
    "endpoint_projection"
  }
}

impl AggregateEvent<EndpointProjection> for SpecEvent {
  fn apply_to(self, aggregate: &mut EndpointProjection) {
    if let SpecEvent::EndpointEvent(event) = self {
      match event {
        EndpointEvent::PathComponentAdded(e) => {
          aggregate.with_path(e.parent_path_id, e.path_id, e.name);
        }
        _ => {}
      }
    }
  }
}
