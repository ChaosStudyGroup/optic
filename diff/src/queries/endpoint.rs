use crate::projections::endpoint::EndpointProjection;
use crate::state::endpoint::PathComponentId;

pub struct EndpointQueries<'a> {
  pub endpoint_projection: &'a EndpointProjection,
}

impl<'a> EndpointQueries<'a> {
  pub fn new(endpoint_projection: &'a EndpointProjection) -> Self {
    EndpointQueries {
      endpoint_projection,
    }
  }
  pub fn resolve_path(&self, path: &String) -> Option<&PathComponentId> {
    // self
    //   .endpoint_projection
    //   .absolute_paths
    //   .iter()
    //   .find_map(|(path_id, absolute_path)| {
    //     if path == absolute_path {
    //       Some(path_id)
    //     } else {
    //       None
    //     }
    //   })
    None
  }
}
