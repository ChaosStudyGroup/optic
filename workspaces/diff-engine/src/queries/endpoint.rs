use crate::projections::endpoint::{Edge, EndpointProjection, Node, ROOT_PATH_ID};
use crate::projections::endpoint::{RequestBodyDescriptor, ResponseBodyDescriptor};
use crate::state::endpoint::{
  HttpMethod, HttpStatusCode, PathComponentId, PathComponentIdRef, RequestId, ResponseId,
};
use crate::HttpInteraction;
use petgraph::graph::Graph;
use petgraph::visit::EdgeFilteredNeighborsDirected;
pub struct EndpointQueries<'a> {
  pub endpoint_projection: &'a EndpointProjection,
}

impl<'a> EndpointQueries<'a> {
  pub fn new(endpoint_projection: &'a EndpointProjection) -> Self {
    EndpointQueries {
      endpoint_projection,
    }
  }
  pub fn resolve_path(&self, interaction: &HttpInteraction) -> Option<PathComponentIdRef> {
    // eprintln!("{}", path);
    let mut path_components = interaction.request.path.split('/');
    // skip leading empty
    path_components.next();
    let mut last_resolved_path_id = Some(ROOT_PATH_ID);
    while let Some(s) = path_components.next() {
      // eprintln!("trying to match segment {}", s);
      let node_index = self.graph_get_index(last_resolved_path_id.unwrap());

      last_resolved_path_id = None;

      // first search for named path components
      let children = self.graph_get_children(node_index);
      for child in children {
        let child_node = self.endpoint_projection.graph.node_weight(child).unwrap();

        match child_node {
          Node::PathComponent(id, descriptor) => {
            // eprintln!("1 - neighbor {}", descriptor.name);
            if !descriptor.is_parameter {
              if descriptor.name == s {
                last_resolved_path_id = Some(id);
                break;
              }
            }
          }
          _ => {}
        }
      }
      if let Some(x) = last_resolved_path_id {
        continue;
      }
      // try path parameters since we didn't find any matches in regular path components
      let children = self.graph_get_children(node_index);
      for child in children {
        let child_node = self.endpoint_projection.graph.node_weight(child).unwrap();
        match child_node {
          Node::PathComponent(id, descriptor) => {
            // eprintln!("2 - neighbor {}", descriptor.name);

            if descriptor.is_parameter {
              last_resolved_path_id = Some(id);
              break;
            }
          }
          _ => {}
        }
      }
      if let Some(x) = last_resolved_path_id {
        continue;
      }
      return None;
    }
    last_resolved_path_id
  }

  pub fn resolve_operations(
    &self,
    interaction: &'a HttpInteraction,
    path_id: PathComponentIdRef,
  ) -> impl Iterator<Item = (&RequestId, &RequestBodyDescriptor)> {
    let path_node_index = self.graph_get_index(path_id);
    let children = self
      .endpoint_projection
      .graph
      .neighbors_directed(*path_node_index, petgraph::Direction::Incoming);
    let matching_method = children
      .filter(move |i| {
        let node = self.endpoint_projection.graph.node_weight(*i).unwrap();
        match node {
          Node::HttpMethod(http_method) => interaction.request.method == *http_method,
          _ => false,
        }
      })
      .flat_map(move |i| {
        let children = self
          .endpoint_projection
          .graph
          .neighbors_directed(i, petgraph::Direction::Incoming);

        let operations = children.filter_map(move |i| {
          let node = self.endpoint_projection.graph.node_weight(i).unwrap();
          match node {
            Node::Request(request_id, request_descriptor) => Some((request_id, request_descriptor)),
            _ => None,
          }
        });
        operations
      });
    matching_method
  }

  pub fn resolve_responses(
    &self,
    interaction: &'a HttpInteraction,
    path_id: PathComponentIdRef,
  ) -> impl Iterator<Item = (&ResponseId, &ResponseBodyDescriptor)> {
    let path_node_index = self.graph_get_index(path_id);
    let children = self
      .endpoint_projection
      .graph
      .neighbors_directed(*path_node_index, petgraph::Direction::Incoming);
    let matching_status_code = children
      .filter(move |i| {
        let node = self.endpoint_projection.graph.node_weight(*i).unwrap();
        // eprintln!("method node {:?}", node);
        match node {
          Node::HttpMethod(http_method) => interaction.request.method == *http_method,
          _ => false,
        }
      })
      .flat_map(move |i| {
        // eprintln!("method node id {:?}", i);
        let children = self
          .endpoint_projection
          .graph
          .neighbors_directed(i, petgraph::Direction::Incoming);
        let status_code_nodes = children.filter_map(move |i| {
          // eprintln!("method child node id {:?}", i);
          let node = self.endpoint_projection.graph.node_weight(i).unwrap();
          match node {
            Node::HttpStatusCode(http_status_code) => {
              // eprintln!("status code {:?}", http_status_code);
              if interaction.response.status_code == *http_status_code {
                Some(i)
              } else {
                None
              }
            }
            _ => None,
          }
        });
        status_code_nodes
      })
      .flat_map(move |i| {
        let children = self
          .endpoint_projection
          .graph
          .neighbors_directed(i, petgraph::Direction::Incoming);
        let response_nodes = children.filter_map(move |i| {
          // eprintln!("status_code child node id {:?}", i);
          let node = self.endpoint_projection.graph.node_weight(i).unwrap();
          // eprintln!("status_code child node {:?}", node);
          match node {
            Node::Response(response_id, response_descriptor) => {
              Some((response_id, response_descriptor))
            }
            _ => None,
          }
        });
        response_nodes
      });
    matching_status_code
  }

  fn graph_get_index(&self, node_id: &str) -> &petgraph::graph::NodeIndex {
    return self
      .endpoint_projection
      .node_id_to_index
      .get(node_id)
      .expect("expected a node with node_id to exist");
  }

  fn graph_get_node(&self, node_id: &str) -> &Node {
    self
      .endpoint_projection
      .graph
      .node_weight(*self.graph_get_index(node_id))
      .expect("expected node with node_id to exist")
  }

  fn graph_get_children(
    &self,
    node_index: &petgraph::graph::NodeIndex,
  ) -> petgraph::graph::Neighbors<Edge> {
    let neighbors = self
      .endpoint_projection
      .graph
      .neighbors_directed(*node_index, petgraph::Direction::Incoming);
    return neighbors;
  }
}
