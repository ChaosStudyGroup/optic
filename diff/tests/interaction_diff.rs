use cqrs_core::Aggregate;
use optic_diff::events::SpecEvent;
use optic_diff::projections::endpoint::EndpointProjection;

#[test]
fn can_yield_diff_results() {
  let events = SpecEvent::from_file(
    std::env::current_dir()
      .unwrap()
      .join("tests/fixtures/ergast-example-spec.json")
      .to_str()
      .unwrap(),
  )
  .expect("example spec should deserialize");

  let mut events_projection = EndpointProjection::default();

  for event in events {
    events_projection.apply(event)
  }
}
