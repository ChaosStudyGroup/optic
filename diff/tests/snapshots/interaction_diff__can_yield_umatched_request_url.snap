---
source: tests/interaction_diff.rs
expression: "Dot::with_config(&endpoint_projection.graph, &[Config::EdgeNoLabel])"
---
digraph {
    0 [ label = "PathComponent(\l    \"root\",\l    PathComponentDescriptor {\l        is_parameter: false,\l        name: \"\",\l    },\l)\l" ]
    1 [ label = "PathComponent(\l    \"path_7K06UY59re\",\l    PathComponentDescriptor {\l        is_parameter: false,\l        name: \"api\",\l    },\l)\l" ]
    2 [ label = "PathComponent(\l    \"path_wv2tYjnRca\",\l    PathComponentDescriptor {\l        is_parameter: false,\l        name: \"f1\",\l    },\l)\l" ]
    3 [ label = "PathComponent(\l    \"path_UslYN0iwbI\",\l    PathComponentDescriptor {\l        is_parameter: true,\l        name: \"season\",\l    },\l)\l" ]
    4 [ label = "HttpMethod(\l    \"GET\",\l)\l" ]
    5 [ label = "Request(\l    \"request_dfwMS3YkPW\",\l    RequestBodyDescriptor {\l        body: None,\l    },\l)\l" ]
    6 [ label = "HttpStatusCode(\l    200,\l)\l" ]
    7 [ label = "Response(\l    \"response_WkmtI23TF7\",\l    ResponseBodyDescriptor {\l        body: Some(\l            BodyDescriptor {\l                http_content_type: \"application/json; charset=utf-8\",\l                root_shape_id: \"EQSZqM_0\",\l            },\l        ),\l    },\l)\l" ]
    8 [ label = "PathComponent(\l    \"path_C3V8NNm66A\",\l    PathComponentDescriptor {\l        is_parameter: true,\l        name: \"round\",\l    },\l)\l" ]
    9 [ label = "PathComponent(\l    \"path_8CU4hvBQlw\",\l    PathComponentDescriptor {\l        is_parameter: false,\l        name: \"constructors\",\l    },\l)\l" ]
    10 [ label = "HttpMethod(\l    \"GET\",\l)\l" ]
    11 [ label = "HttpStatusCode(\l    200,\l)\l" ]
    12 [ label = "Response(\l    \"response_aFeAkBC7kd\",\l    ResponseBodyDescriptor {\l        body: Some(\l            BodyDescriptor {\l                http_content_type: \"application/json; charset=utf-8\",\l                root_shape_id: \"RdXJj0_0\",\l            },\l        ),\l    },\l)\l" ]
    13 [ label = "Request(\l    \"request_4toesS0Kno\",\l    RequestBodyDescriptor {\l        body: None,\l    },\l)\l" ]
    14 [ label = "PathComponent(\l    \"path_NbecIObO61\",\l    PathComponentDescriptor {\l        is_parameter: false,\l        name: \"drivers\",\l    },\l)\l" ]
    15 [ label = "HttpMethod(\l    \"GET\",\l)\l" ]
    16 [ label = "HttpStatusCode(\l    200,\l)\l" ]
    17 [ label = "Response(\l    \"response_Hi7nMNuQMO\",\l    ResponseBodyDescriptor {\l        body: Some(\l            BodyDescriptor {\l                http_content_type: \"application/json; charset=utf-8\",\l                root_shape_id: \"9zYywG_0\",\l            },\l        ),\l    },\l)\l" ]
    18 [ label = "Request(\l    \"request_gsOGDZpWvZ\",\l    RequestBodyDescriptor {\l        body: None,\l    },\l)\l" ]
    19 [ label = "PathComponent(\l    \"path_BkUCPF4bFA\",\l    PathComponentDescriptor {\l        is_parameter: false,\l        name: \"results\",\l    },\l)\l" ]
    20 [ label = "HttpMethod(\l    \"GET\",\l)\l" ]
    21 [ label = "HttpStatusCode(\l    200,\l)\l" ]
    22 [ label = "Response(\l    \"response_7ZwwiBUQ2h\",\l    ResponseBodyDescriptor {\l        body: Some(\l            BodyDescriptor {\l                http_content_type: \"application/json; charset=utf-8\",\l                root_shape_id: \"Om3KlN_0\",\l            },\l        ),\l    },\l)\l" ]
    23 [ label = "Request(\l    \"request_dxZmzIoGI7\",\l    RequestBodyDescriptor {\l        body: None,\l    },\l)\l" ]
    24 [ label = "HttpMethod(\l    \"GET\",\l)\l" ]
    25 [ label = "HttpStatusCode(\l    200,\l)\l" ]
    26 [ label = "Response(\l    \"response_mLWnPnRfdJ\",\l    ResponseBodyDescriptor {\l        body: Some(\l            BodyDescriptor {\l                http_content_type: \"application/json; charset=utf-8\",\l                root_shape_id: \"1Uivbu_0\",\l            },\l        ),\l    },\l)\l" ]
    27 [ label = "Request(\l    \"request_nUhlkPiAEN\",\l    RequestBodyDescriptor {\l        body: None,\l    },\l)\l" ]
    28 [ label = "PathComponent(\l    \"path_lR5K6vJW0u\",\l    PathComponentDescriptor {\l        is_parameter: false,\l        name: \"drivers\",\l    },\l)\l" ]
    29 [ label = "PathComponent(\l    \"path_P9esbhMNb0\",\l    PathComponentDescriptor {\l        is_parameter: true,\l        name: \"driverId\",\l    },\l)\l" ]
    30 [ label = "HttpMethod(\l    \"GET\",\l)\l" ]
    31 [ label = "HttpStatusCode(\l    200,\l)\l" ]
    32 [ label = "Response(\l    \"response_mByFnYAel4\",\l    ResponseBodyDescriptor {\l        body: Some(\l            BodyDescriptor {\l                http_content_type: \"application/json; charset=utf-8\",\l                root_shape_id: \"TO7LE6_0\",\l            },\l        ),\l    },\l)\l" ]
    33 [ label = "Request(\l    \"request_WQYLPrDfdA\",\l    RequestBodyDescriptor {\l        body: None,\l    },\l)\l" ]
    34 [ label = "PathComponent(\l    \"path_2iEuAjriND\",\l    PathComponentDescriptor {\l        is_parameter: false,\l        name: \"following\",\l    },\l)\l" ]
    35 [ label = "PathComponent(\l    \"path_MwY9nZ6x25\",\l    PathComponentDescriptor {\l        is_parameter: false,\l        name: \"drivers\",\l    },\l)\l" ]
    36 [ label = "HttpMethod(\l    \"POST\",\l)\l" ]
    37 [ label = "Request(\l    \"request_NR43nZPaOr\",\l    RequestBodyDescriptor {\l        body: Some(\l            BodyDescriptor {\l                http_content_type: \"application/json\",\l                root_shape_id: \"SGyna3_0\",\l            },\l        ),\l    },\l)\l" ]
    38 [ label = "HttpStatusCode(\l    201,\l)\l" ]
    39 [ label = "Response(\l    \"response_bK58ebkIt4\",\l    ResponseBodyDescriptor {\l        body: Some(\l            BodyDescriptor {\l                http_content_type: \"application/json\",\l                root_shape_id: \"Dzhqpq_0\",\l            },\l        ),\l    },\l)\l" ]
    40 [ label = "HttpStatusCode(\l    400,\l)\l" ]
    41 [ label = "Response(\l    \"response_ziGFJv9b29\",\l    ResponseBodyDescriptor {\l        body: Some(\l            BodyDescriptor {\l                http_content_type: \"application/json\",\l                root_shape_id: \"shape_lQpti9SvC2\",\l            },\l        ),\l    },\l)\l" ]
    42 [ label = "Response(\l    \"response_oKNTHijcIc\",\l    ResponseBodyDescriptor {\l        body: Some(\l            BodyDescriptor {\l                http_content_type: \"application/json\",\l                root_shape_id: \"shape_hqZddtlZ7U\",\l            },\l        ),\l    },\l)\l" ]
    43 [ label = "Response(\l    \"response_nxWT5qcYhF\",\l    ResponseBodyDescriptor {\l        body: Some(\l            BodyDescriptor {\l                http_content_type: \"application/json\",\l                root_shape_id: \"shape_VGtg2HMvjl\",\l            },\l        ),\l    },\l)\l" ]
    1 -> 0 [ ]
    2 -> 1 [ ]
    3 -> 2 [ ]
    4 -> 3 [ ]
    5 -> 4 [ ]
    6 -> 4 [ ]
    7 -> 6 [ ]
    8 -> 3 [ ]
    9 -> 8 [ ]
    10 -> 9 [ ]
    11 -> 10 [ ]
    12 -> 11 [ ]
    13 -> 10 [ ]
    14 -> 8 [ ]
    15 -> 14 [ ]
    16 -> 15 [ ]
    17 -> 16 [ ]
    18 -> 15 [ ]
    19 -> 8 [ ]
    20 -> 19 [ ]
    21 -> 20 [ ]
    22 -> 21 [ ]
    23 -> 20 [ ]
    24 -> 8 [ ]
    25 -> 24 [ ]
    26 -> 25 [ ]
    27 -> 24 [ ]
    28 -> 3 [ ]
    29 -> 28 [ ]
    30 -> 29 [ ]
    31 -> 30 [ ]
    32 -> 31 [ ]
    33 -> 30 [ ]
    34 -> 0 [ ]
    35 -> 34 [ ]
    36 -> 35 [ ]
    37 -> 36 [ ]
    38 -> 36 [ ]
    39 -> 38 [ ]
    40 -> 36 [ ]
    41 -> 40 [ ]
    42 -> 31 [ ]
    43 -> 21 [ ]
}

